TITLE: Comment résoudre un problème d&#39;accès à la configuration du service LDAP
AUTHOR: Philippe Latu
DATE: dimanche 23 septembre 2012, 17:24:57 (UTC+0200)
DESC: 
FORMAT: raw
-----
BODY:
<p>En révisant le support de travaux pratiques sur <a
href='http://www.inetdoc.net/travaux_pratiques/index.html#sysadm-net.ldap'>l'introduction
aux annuaires LDAP avec OpenLDAP</a>, j'ai perdu <i>un certain temps</i> avant
de pouvoir traiter la partie sur <em>l'analyse de la configuration du service
LDAP</em>. Relativement à l'édition précédente, les versions du paquet source
<tt>openldap</tt> ont évolué et un petit détail sur la résolution des noms
m'avait échappé. Bien sûr, me direz-vous, dans la vraie vie tout le monde
installe son service LDAP en ayant préalablement validé que le serveur est
connu du service DNS. Eh bien, pas moi ! Dans un contexte de travaux pratiques,
on cherche à isoler les services les uns des autres de façon à ne traiter qu'un
seul problème bien identifié à la fois. Voyons donc comment on peut se
fourvoyer facilement et quelle démarche aurait pu être suivie d'entrée de
jeu.</p>

<h3>Au début, il était question d'authentification ...</h3>

<p>Pour valider les questions et les réponses du TP, on configure de façon très
classique deux instances de machines virtuelles avec le système de base et le
service SSH en <i>Debian/testing</i> : un serveur et un client. On effectue les
mises à jour et on renomme les machines. On obtient avec la topologie
suivante :</p>

<pre>          |eth0
         _|____
       (_______)           noyau
       |Routeur|    <-- système hôte
       (_______)
          |br0  @IP: 198.51.100.1
       ___|______
      |          |  <-- openvswitch
      `-.------.-'
	|tap0  |tap1
        |    --'.
        |    |' |hostname : vm-ldap-client
        |    |. |@IP      : 198.51.100.3
      --'.   `--'
      |' |hostname : <b><em>vm-ldap-server</em></b>
      |. |@IP      : 198.51.100.2
      `--'</pre>

<p>Après installation des deux paquets <tt>slapd</tt> et <tt>ldap-utils</tt>
sur le serveur, on teste l'accès à la configuration.</p>

<pre>Traitement des actions différées (« triggers ») pour « man-db »...
Paramétrage de libltdl7:amd64 (2.4.2-1.1) ...
Paramétrage de libodbc1:amd64 (2.2.14p2-5) ...
Paramétrage de libperl5.14 (5.14.2-13) ...
Paramétrage de libslp1 (1.2.1-9) ...
Paramétrage de slapd (2.4.31-1) ...
  Creating new user openldap... done.
  Creating initial configuration... done.
  Creating LDAP directory... done.
[ ok ] Starting OpenLDAP: slapd.
Paramétrage de ldap-utils (2.4.31-1) ...
localepurge: Disk space freed in /usr/share/locale: 0 KiB
localepurge: Disk space freed in /usr/share/man: 0 KiB

Total disk space freed by localepurge: 0 KiB

# ldapsearch -Y EXTERNAL -H ldapi:/// -b "cn=config" dn
<b style="color: red;">ldap_sasl_interactive_bind_s: Local error (-2)</b></pre>

<p>Eh paf ! C'est à partir de la que les choses ont mal tourné. Je me suis
focalisé sur l'authentification SASL et non sur la partie
<tt>interactive_bind_s</tt> du message. Pourtant, cette instruction est bien
celle préconisée dans le fichier de documentation du paquet <tt>slapd</tt> :
<tt>/usr/share/doc/slapd/README.Debian.gz</tt></p>

<p>Je me suis lancé dans la configuration de l'authentification SASL avec le
paquet <tt>sasl2-bin</tt> et le mécanisme PAM pour authentifier les
utilisateurs locaux dont le super-utilisateur qui doit avoir accès sans
restriction à la configuration du ou des annuaires LDAP configurés sur le
serveur. Bien que cette authentification soit fonctionnelle, j'obtenais
toujours le même message d'erreur. La situation était bloquée. Pas moyen de
modifier le DIT.</p>

<h3>À la fin, il n'était plus question que de résolution des noms</h3>

<p>Au lieu de m'enfermer dans cette voie sans issue, j'aurais dû commencer par
faire appel à <tt>strace</tt>.</p>

<pre># strace ldapsearch -Y EXTERNAL -H ldapi:/// -b "cn=config" dn
<i>&lt;snipped></i>
recvfrom(4, "h4\201\203\0\1\0\0\0\1\0\0\16vm-ldap-server\0\0\34\0\1"..., 2048, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("198.51.100.1")}, [16]) = 107
gettimeofday({1348409171, 770189}, NULL) = 0
poll([{fd=4, events=POLLIN}], 1, 4998)  = 1 ([{fd=4, revents=POLLIN}])
ioctl(4, FIONREAD, [107])               = 0
recvfrom(4, "6\361\201\203\0\1\0\0\0\1\0\0\16vm-ldap-server\0\0\1\0\1"..., 1941, 0, {sa_family=AF_INET, sin_port=htons(53), sin_addr=inet_addr("198.51.100.1")}, [16]) = 107
close(4)                                = 0
write(2, "ldap_sasl_interactive_bind_s: Lo"..., 47<b style="color: red;">ldap_sasl_interactive_bind_s: Local error (-2)</b>
) = 47
write(3, "0\5\2\1\1B\0", 7)             = 7
close(3)</pre>

<p>Voici à nouveau le message d'erreur consécutivement à une transaction réseau
avec des sockets. Mais que diable vient faire cette requête réseau dans une
manipulation locale au système ? On passe donc à l'analyse réseau sur le
système hôte ; celui avec l'adresse IP 198.51.100.1.</p>

<pre># tshark -i br0 ! port 22
tshark: Lua: Error during loading:
 [string "/usr/share/wireshark/init.lua"]:45: dofile has been disabled
Running as user "root" and group "root". This could be dangerous.
Capturing on br0
  0.000000 198.51.100.2 -> 198.51.100.1 DNS 74 Standard query 0x6213  A vm-ldap-server
  0.000266 198.51.100.1 -> 198.51.100.2 DNS 149 Standard query response 0x6213 No such name
  0.016816 198.51.100.2 -> 198.51.100.1 DNS 74 Standard query 0x2cd4  A vm-ldap-server
  0.016885 198.51.100.2 -> 198.51.100.1 DNS 74 Standard query 0x28a4  AAAA vm-ldap-server
  0.017050 198.51.100.1 -> 198.51.100.2 DNS 149 Standard query response 0x28a4 No such name
  0.017110 198.51.100.1 -> 198.51.100.2 DNS 149 Standard query response 0x2cd4 No such name</pre>

<p>Bingo ! la requête locale sur le DIT de l'annuaire à besoin de connaître la
correspondance entre nom et adresse IP. On se jette sur l'édition du fichier
<tt>/etc/hosts</tt> dans lequel on donne l'information manquante.</p>

<pre># head -4 /etc/hosts
127.0.0.1       localhost
127.0.1.1       vm-ldap-server

198.51.100.2    vm-ldap-server</pre>

<p>Ouf ! Une nouvelle exécution de la commande <tt>ldapsearch</tt> ne produit
plus d'erreur et on peut enfin modifier le DIT de l'annuaire.</p>

<pre># ldapsearch -Y EXTERNAL -H ldapi:/// -b "cn=config" dn
SASL/EXTERNAL authentication started
SASL username: gidNumber=0+uidNumber=0,cn=peercred,cn=external,cn=auth
SASL SSF: 0
# extended LDIF
#
# LDAPv3
# base &lt;cn=config> with scope subtree
# filter: (objectclass=*)
# requesting: dn 
#
<i>&lt;snipped></i>
# search result
search: 2
result: 0 Success

# numResponses: 12
# numEntries: 11</pre>

<h3>En guise de conclusion</h3>

<p>C'est une nouvelle petite leçon d'humilité. On s'imagine être expérimenté et
on se fait attraper comme un débutant sur un détail insignifiant en
apparence.</p>

<p>Maintenant, pour rebondir sur le billet intitulé <a
href='http://raphaelhertzog.fr/go/bugreporting/'>7 astuces pour rapporter les
bogues Debian efficacement et voir vos problèmes résolus</a>, est-ce que le
problème que je viens de rencontrer mérite un rapport de bug ? Il m'est déjà
arrivé dans le passé d'émettre des rapports qui ont fait flop. Ils étaient soit
hors sujet (#616487), soit sans aucun intérêt.</p>

END-----
